lib/msf/base/sessions/command_shell.rb

```diff
@@ -312,18 +312,18 @@ def cmd_shell(*args)
   end
 
   #
-  # Check if there is a binary in PATH env
+  # Returns path of a binary in PATH env.
   #
   def binary_exists(binary)
     print_status("Trying to find binary(#{binary}) on target machine")
-    binary_path = shell_command_token("which #{binary}").to_s.strip
-    if binary_path.eql?("#{binary} not found") || binary_path == ""
-      print_error(binary_path)
+    binary_path = shell_command_token("command -v '#{binary}' && echo true")
+    if binary_path.to_s.eql?("")
+      print_error("#{binary} not found")
```

Comment:
Instead of printing the error straight away I think we should fall back to `which` here to support non-POSIX compliant shells, like ESXi and older busybox.

Reply:
okay, Will change that :)

Reply:
> Instead of printing the error straight away I think we should fall back to `which` here

`which` is error prone. This reintroduces the problem that this PR is trying to solve.

> to support non-POSIX compliant shells

Non-POSIX *nix seems unlikely. 

> like ESXi and older busybox.

Is this an observation or a guess?

Reply:
> Is this an observation or a guess?

From an ESXi box:
```
[user@esxi:~] command
-sh: command: not found
[user@esxi:~] which
BusyBox v1.22.1 (2016-03-26 07:03:26 PDT) multi-call binary.Usage: which [COMMAND]...Locate a COMMAND
```

Reply:
I thought there is not harm in falling to `which` if `command` fails, so added this. Please let me know if it needs to be reverted or some other changes are required. XD